src/vdaf/mastic.rs

```diff
@@ -0,0 +1,727 @@
+        if self.result.len() != output_share.result.len() {
+            return Err(VdafError::Uncategorized(
+                "Attempted to accumulate two output shares with different agg_params".to_string(),
+            ));
+        };
+        // Would love to get rid of the below clone if possible.
+        for (a, o) in self.result.iter_mut().zip(output_share.result.iter()) {
+            *a += o.clone();
+        }
+        Ok(())
+    }
+}
+
+impl<'a, T, P, const SEED_SIZE: usize>
+    ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, &'a MasticAggregationParam)>
+    for MasticOutputShare<VidpfWeight<T::Field>>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        decoding_parameter: &(&Mastic<T, P, SEED_SIZE>, &MasticAggregationParam),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let (mastic, agg_param) = decoding_parameter;
+        let l = agg_param.prefixes().len();
+        let mut result = Vec::<VidpfWeight<T::Field>>::with_capacity(l);
+        for _ in 0..l {
+            result.push(VidpfWeight::<T::Field>::decode_with_param(
+                mastic.vidpf().weight_parameter(),
+                bytes,
+            )?);
+        }
+        Ok(MasticOutputShare { result })
+    }
+}
+
+impl<V: VidpfValue> Encode for MasticOutputShare<V> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        for elem in &self.result {
+            elem.encode(bytes)?;
+        }
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        let mut total = 0;
+        for elem in &self.result {
+            total += elem.encoded_len()?
+        }
+        Some(total)
+    }
+}
+
+impl<V: VidpfValue> PartialEq for MasticOutputShare<V> {
+    fn eq(&self, other: &MasticOutputShare<V>) -> bool {
+        self.result == other.result
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Vdaf for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type Measurement = (VidpfInput, T::Measurement);
+    type AggregateResult = T::AggregateResult;
+    type AggregationParam = MasticAggregationParam;
+    type PublicShare = MasticPublicShare<VidpfWeight<T::Field>>;
+    type InputShare = MasticInputShare<T::Field, SEED_SIZE>;
+    type OutputShare = MasticOutputShare<VidpfWeight<T::Field>>;
+    type AggregateShare = MasticOutputShare<VidpfWeight<T::Field>>;
+
+    fn algorithm_id(&self) -> u32 {
+        self.algorithm_id
+    }
+
+    fn num_aggregators(&self) -> usize {
+        2
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn shard_with_random(
+        &self,
+        measurement_label: &VidpfInput,
+        measurement_weight: &VidpfWeight<T::Field>,
+        nonce: &[u8; 16],
+        vidpf_keys: [VidpfKey; 2],
+        szk_random: [Seed<SEED_SIZE>; 2],
+        opt_random: Option<Seed<SEED_SIZE>>,
```

Comment:
"opt_random" is a somewhat ambiguous term in this context: we have some randomness for the SZK; what is an "OPT"?